File: rating-service/src/main/java/com/example/rating_service/controller/kafka/UserCreatedEventListener.java
package com.example.rating_service.controller.kafka;

import com.example.rating_service.model.events.UserCreatedEvent;
import com.example.rating_service.service.RatingService;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.kafka.annotation.KafkaListener;
import org.springframework.messaging.handler.annotation.Header;
import org.springframework.stereotype.Component;

@RequiredArgsConstructor
@Component
@Slf4j
public class UserCreatedEventListener {
    private final ObjectMapper objectMapper;
    private final RatingService service;

    @Value("${spring.kafka.event.user-created}")
    private String userCreatedEventName;

    @KafkaListener(topics = "${spring.kafka.topic.rides.users}")
    public void listen(String payload, @Header("eventType") String eventTypeString) {
        try {
            log.info("Received {} event", eventTypeString);
            if (eventTypeString.equals(userCreatedEventName)) {
                var event = objectMapper.readValue(payload, UserCreatedEvent.class);
                service.createUser(event.userId());
            }
        } catch (JsonProcessingException e) {
            log.error("Json converting exception: {}", e.getMessage());
            throw new RuntimeException("Json converting exception", e);
        } catch (IllegalArgumentException e) {
            throw new IllegalArgumentException("Unknown event type: " + eventTypeString);
        }
    }
}

File: driver-service/src/main/kotlin/com/example/driver_service/controller/listener/UserCreatedEventListener.kt
package com.example.driver_service.controller.listener

import com.example.driver_service.model.event.UserRegisteredEvent
import com.example.driver_service.service.DriverService
import com.fasterxml.jackson.databind.ObjectMapper
import mu.KotlinLogging
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.kafka.annotation.KafkaListener
import org.springframework.messaging.handler.annotation.Header
import org.springframework.messaging.handler.annotation.Payload
import org.springframework.stereotype.Component

@Component
class UserCreatedEventListener(
    private val objectMapper: ObjectMapper,
    @Qualifier("kafkaTypeMapping")
    private val typeMapping: Map<String, Class<out Any>>,
    private val service: DriverService
) {
    companion object {
        private val log = KotlinLogging.logger {}
    }

    @KafkaListener(topics = ["\${spring.kafka.topic.rides.users}"])
    fun listen(
        @Payload payload: String,
        @Header("eventType") type: String
    ) {
        try {
            log.info { "Processing event: $type" }
            val targetClass = typeMapping[type]
                ?: throw IllegalArgumentException("Unknown event type: $type")

            when (val event = objectMapper.readValue(payload, targetClass)) {
                is UserRegisteredEvent -> service.save(event)
                else -> log.warn { "No handler found for class: ${targetClass.simpleName}" }
            }
        } catch (e: Exception) {
            log.error(e) { "Error while processing kafka message: ${e.message} in driver searching listener!" }
        }
    }
}